androidbrowserhelper/src/main/java/com/google/androidbrowserhelper/trusted/DelegationService.java

```diff
@@ -14,19 +14,145 @@
+
+    private static final String CHECK_LOCATION_PERMISSION_COMMAND_NAME = "checkAndroidLocationPermission";
+    private static final String LOCATION_PERMISSION_RESULT = "locationPermissionResult";
+
+    private static final String START_LOCATION_COMMAND_NAME = "startLocation";
+    private static final String STOP_LOCATION_COMMAND_NAME = "stopLocation";
+
+    private static final String EXTRA_RESULT_RECEIVER = "EXTRA_RESULT_RECEIVER";
+    private static final String EXTRA_PERMISSIONS = "EXTRA_PERMISSIONS";
+    private static final String EXTRA_GRANT_RESULTS = "EXTRA_GRANTED_RESULT";
+
+    private static final String mLocationPermission = Manifest.permission.ACCESS_FINE_LOCATION;
+
+    private boolean mIsProviderGmsCore;
 
     @NonNull
     @Override
     public TokenStore getTokenStore() {
         return new SharedPreferencesTokenStore(this);
     }
+
+    @Nullable
+    @Override
+    public Bundle onExtraCommand(
+            String commandName, Bundle args, TrustedWebActivityCallbackRemote callback) {
+        Log.d(TAG, commandName);
+        Bundle result = new Bundle();
+        switch (commandName) {
+            case CHECK_LOCATION_PERMISSION_COMMAND_NAME: {
+                if (callback == null) break;
+
+                result.putBoolean(commandName, true);
+                requestPermission(callback, mLocationPermission);
+                break;
+            }
+            case START_LOCATION_COMMAND_NAME: {
+                if (callback == null) return null;
+                if (!checkPermission(mLocationPermission)) return null;
+                startLocationProvider(callback, args.getBoolean("enableHighAccuracy"));
+
+                result.putBoolean(commandName, true);
+                break;
+            }
+            case STOP_LOCATION_COMMAND_NAME: {
+                stopLocationProvider();
+
+                result.putBoolean(commandName, true);
+                break;
+            }
+            default: {
+                result.putBoolean(commandName, false);
+            }
+        }
+        return result;
+    }
+
+    private boolean checkPermission(String permission) {
+        return ContextCompat.checkSelfPermission(this, permission) == PackageManager.PERMISSION_GRANTED;
+    }
+
+    private void requestPermission(TrustedWebActivityCallbackRemote callback, String permission) {
+        if (checkPermission(permission)) {
+            notifyPermissionResult(callback,
+                    new String[]{permission}, new int[]{PackageManager.PERMISSION_GRANTED});
+            return;
+        }
+
+        Handler handler = new Handler(Looper.getMainLooper(), message -> {
+            Bundle data = message.getData();
+            notifyPermissionResult(callback,
+                    data.getStringArray(EXTRA_PERMISSIONS), data.getIntArray(EXTRA_GRANT_RESULTS));
+            return true;
+        });
+        final Messenger messenger = new Messenger(handler);
+        handler.post(() -> {
```

Comment:
Do you need to do this? Can you not call startActivity from a background thread?